### ISSUE TITLE
Using --no-codegen blocks --emit options

### ISSUE THREAD

**Issue by bew** [crystal-lang/crystal#5820]
When we do: `crystal build hello.cr --emit llvm-ir --no-codegen`
It'll not generate the `hello.ll` file even though it was asked for.

This is because the emit options are handled all-in-one-place in
https://github.com/crystal-lang/crystal/blob/80cbe6603f938fc2200b785bd3fb2edbd854a17d/src/compiler/crystal/compiler.cr#L676-L687

Which is called in the `codegen` block:
https://github.com/crystal-lang/crystal/blob/80cbe6603f938fc2200b785bd3fb2edbd854a17d/src/compiler/crystal/compiler.cr#L339-L353
At line 348.

I think it should emit what was asked for as soon as it get it, so the llvm-ir after it was built, the obj in the codegen phase, etc..

**Comment by bew** [comment]
This also means that there is no way to dump llvm IR when there is a `module validation failed` error, that we would like to debug.

**Comment by straight-shoota** [comment]
In order to debug #5972 I had to modify the compiler manually to disable module validation. So +1 to that!

**Comment by bew** [comment]
While looking into that, I noticed that all the codegen phases (mainly crystal to LLVM IR and LLVM to BC+OBJ) are done in the `codegen` method.
I would suggest to separate the Crystal to LLVM IR codegen phase from the binary codegen (LLVM-IR to BC+OBJ), maybe named `IR codegen` ?

Also the current different emit options are in 2 categories:
- after llvm ir generation: for `llvm-ir` (should be processed even on `--no-codegen`)
- after binary codegen: for `asm`, `llvm-bc`, `obj` (not processed on `--no-codegen`)

Those are currently all handled and processed after binary codegen. By having the `IR codegen` separated, we could handle the different emit options at different time.

WDYT?

**Comment by asterite** [comment]
What's so bad about not using `--no-codegen` when using `--emit`? No codegen means the codegen phase doesn't run, and this trivially means no artifacts (.ll, .s, .o, final executable) will be produced.

**Comment by straight-shoota** [comment]
@asterite True, these are all different stages of codegen. But what do you do if you want the `.ll` (or `.s`) but not `.o`?

`--no-codegen --emit llvm-ir` makes sense for this: It essentially says "skip codegen, but emit LLVM-IR" which can be interpreted as "do codegen but abort after LLVM-IR is dumped".

**Comment by asterite** [comment]
> True, these are all different stages of codegen. But what do you do if you want the .ll (or .s) but not .o?

You just wait a little more? :-)

> It essentially says "skip codegen, but emit LLVM-IR" which can be interpreted as "do codegen but abort after LLVM-IR is dumped".

There's the confusion. Codegen means "run code to create the LLVM in memory". Emit comes **after** that. Of course we could change that, but I don't see the point. Just wait a few more seconds and you'll have it.

**Comment by asterite** [comment]
Or, well, put another one, if someone wants to implement it, please send a PR (I won't)

**Comment by bew** [comment]
Also, there could be a `--no-binary-codegen` in addition to `--no-codegen`, where the latter disables all codegen (ir & binary), and the former disable the binary codegen only, but allow the ir codegen to be done.

I think it allows more control, and remove _weird_ edge cases like "always do IR codegen when --emit is llvm-ir even on --no-codegen".

**Comment by mattrberry** [comment]
Agree with HertzDevil above. It would be useful to be able to emit and inspect the IR without needing to link
